pkg/controller/mcmhub/hook.go

```diff
@@ -0,0 +1,289 @@
+
+	//SetSuffixFunc let user reset the suffixFunc rule of generating the suffix
+	//of hook instance name
+	SetSuffixFunc(SuffixFunc)
+	//ApplyPreHook returns a type.NamespacedName of the preHook
+	ApplyPreHook(types.NamespacedName) (types.NamespacedName, error)
+	IsPreHookCompleted(types.NamespacedName) (bool, error)
+	ApplyPostHook(types.NamespacedName) (types.NamespacedName, error)
+	IsPostHookCompleted(types.NamespacedName) (bool, error)
+	IsSubscriptionCompleted(types.NamespacedName) (bool, error)
+}
+
+type Hooks struct {
+	pre  unstructured.Unstructured
+	post unstructured.Unstructured
+	//store all the applied prehook instance
+	preHooks  []unstructured.Unstructured
+	postHooks []unstructured.Unstructured
+
+	//store last subscription instance used for the hook operation
+	lastSub *subv1.Subscription
+}
+
+type AnsibleHooks struct {
+	gitClt GitOps
+	clt    client.Client
+	// subscription namespacedName will points to hooks
+	registry   map[types.NamespacedName]*Hooks
+	suffixFunc SuffixFunc
+	//logger
+	logger logr.Logger
+}
+
+// make sure the AnsibleHooks implementate the HookProcessor
+var _ HookProcessor = &AnsibleHooks{}
+
+func NewAnsibleHooks(clt client.Client, logger logr.Logger) *AnsibleHooks {
+	if logger == nil {
+		logger = klogr.New()
+		logger.WithName("ansiblehook")
+	}
+
+	return &AnsibleHooks{
+		clt:        clt,
+		gitClt:     NewHookGit(clt, logger),
+		registry:   map[types.NamespacedName]*Hooks{},
+		logger:     logger,
+		suffixFunc: suffixFromUUID,
+	}
+}
+
+func (a *AnsibleHooks) SetSuffixFunc(f SuffixFunc) {
+	if f == nil {
+		return
+	}
+
+	a.suffixFunc = f
+}
+
+func (a *AnsibleHooks) DegisterSubscription(subKey types.NamespacedName) error {
+	delete(a.registry, subKey)
+	return nil
+}
+
+func (a *AnsibleHooks) RegisterSubscription(subKey types.NamespacedName) error {
+	a.logger.V(DebugLog).Info("entry register subscription")
+	defer a.logger.V(DebugLog).Info("exit register subscription")
+
+	subIns := &subv1.Subscription{}
+	if err := a.clt.Get(context.TODO(), subKey, subIns); err != nil {
+		// subscription is deleted
+		if kerr.IsNotFound(err) {
+			return nil
+		}
+
+		return err
+	}
+
+	preHook := subIns.Spec.Prehook
+	postHook := subIns.Spec.Posthook
+
+	if len(preHook) == 0 && len(postHook) == 0 {
+		return nil
+	}
+
+	//check if the subIns have being changed compare to the hook registry
+	if !a.isUpdateSubscription(subIns) {
+		return nil
+	}
+
+	jobs, err := a.gitClt.DownloadHookResource(subIns)
+
+	if err != nil {
+		return err
+	}
+
+	printJobs(jobs, a.logger)
+	//update the base Ansible job and append a generated job to the preHooks
+	return a.addHookToResitry(subIns, jobs)
+}
+
+func printJobs(jobs []unstructured.Unstructured, logger logr.Logger) {
+	for _, job := range jobs {
+		logger.V(3).Info(fmt.Sprintf("download jobs %#v", job))
+	}
+}
+
+type SuffixFunc func(*subv1.Subscription) string
+
+func suffixFromUUID(subIns *subv1.Subscription) string {
```

Comment:
`subscription.meta.uid` will not change when the subscription is updated. So if the main subscription is updated, it requires to launch a new pre-/post- ansible job with unique name. If reply on uid, the  suffixFromUUID function will generate the same suffix. That will cause the ansible job with same name updated. no new ansible job is created. 

Notice by current ansible job opeator implementation, only when a new ansible job CR is created, a new ansible tower job will be launched.  So from subscription repo,  we should make sure whenever the main subscription is updated, a new ansible job CR should be created.  

May consider to apply `subscription.spec.resourceVersion` for suffix generation. Its value changes when the subscription is updated.

Reply:
will use the `spec.resourceVersion`